src/main/java/MealDemo/service/MealPlanningServiceImpl.java

```diff
@@ -82,4 +80,84 @@ public void updateMeal(int mealId, Meal meal) {
         System.out.println("meal has been updated. Meal details \n" + existingMeal.toString());
     }
 
-}
\ No newline at end of file
+
+    @Override
+    public Map<String, List<Meal>> mealsByCategories() {
+        List<Meal> allMeals = getMeals();
+        Map<String, List<Meal>> mealsByCategoriesList = new HashMap<>();
+        for (Meal meal : allMeals) {
+            mealsByCategoriesList.computeIfAbsent(meal.getCategory(), k -> new ArrayList<>()).add(meal);
+        }
+
+        return mealsByCategoriesList;
+    }
+
+
+    private int randomNumberGenerator(int max) {
+        Random random = new Random();
+        return random.nextInt(max);
+    }
+
+
+    @Override
+    // generate 7 random meals per each day
+    public HashMap<DaysOfTheWeek, Meal> generateWeeklyMeals(HashMap<String, Integer> userInput) {
+        HashMap<DaysOfTheWeek, Meal> weeklyPlan = new HashMap<>();
+        Map<String, List<Meal>> mealsByCategories = mealsByCategories();
+
+        HashMap<String, Integer> request = new HashMap<>(userInput);
+        if (request.isEmpty()) {
+            request.put("Chicken", 2);
+            request.put("Fish", 1);
+            request.put("Meat", 2);
+            request.put("Vegetarian", 2);
+        }
+
+        for (DaysOfTheWeek day : DaysOfTheWeek.values()) {
+            String[] categories = request.keySet().toArray(new String[request.size()]);
```

Comment:
A slight optimization: Move this line to outside of the for loop. Then inside this if check:

```
if (requestNumber == 0) {
                request.remove(randomCategory);
            } else {
                request.replace(randomCategory, requestNumber);
            }
```

Change it to be this:

```
           if (requestNumber == 0) {
                request.remove(randomCategory);
                categories = request.keySet().toArray(new String[request.size()]);
            } else {
                request.replace(randomCategory, requestNumber);
            }
```

You see currently you regenerating this array for EVERY iteration of the loop, you don't need to. You should only change it when you NEED to change it and that only happens when a catergory dies.

Reply:
that makes so much sense!